kong/core/reports.lua

```diff
@@ -76,56 +76,57 @@ local function send(t, host, port)
   local sock = udp_sock()
   local ok, err = sock:setpeername(host, port)
   if not ok then
-    log_error("could not set peer name for UDP socket: ", err)
+    log(ERR, "could not set peer name for UDP socket: ", err)
     return
   end
 
   sock:settimeout(1000)
 
   ok, err = sock:send("<14>"..msg) -- syslog facility code 'log alert'
   if not ok then
-    log_error("could not send data: ", err)
+    log(ERR, "could not send data: ", err)
   end
 
   ok, err = sock:close()
   if not ok then
-    log_error("could not close socket: ", err)
+    log(ERR, "could not close socket: ", err)
   end
 end
 
 ---------------
 -- ping handler
 ---------------
 
+local function get_lock()
+  -- the lock is held for the whole interval to prevent multiple
+  -- worker processes from sending the test request simultaneously.
+  -- here we substract the lock expiration time by 1ms to prevent
+  -- a race condition with the next timer event.
+  return cache.rawadd(KEY, true, PING_INTERVAL - 0.001)
```

Comment:
We should use the new `ngx.shared.kong` dict here. It is meant especially for long-lived data that shouldn't get deleted in an LRU fashion (unlike the `cache` dict which is its purpose).

(Only write to `ngx.shared.kong` in a safe way with `safe_set` and `safe_add` as we want to report `no memory` errors to the user in that case, so it can be increased).